### PR TITLE
feat(rpc): Cookie auth system for the RPC endpoint

### LINKED ISSUE #8864: Protect the Zebra RPC endpoint
The Zebra RPC endpoint is currently unencrypted and accessible by anyone who knows the address and port. While the current recommendation is to restrict access to localhost or a trusted local network, there are cases where the endpoint needs to be open to the internet in production environments.

Implement one of the following authentication methods:

- Basic HTTP Authentication: This method would require users to provide a username and password when connecting to the RPC endpoint.
- Cookie-based Authentication: Similar to Zcashd, this method uses a randomly generated cookie stored on the file system to authenticate RPC connections.

We could implement just one of these methods or explore alternative solutions.  

Authentication should be enabled by default but turning it off should be an option for testing purposes and setups where the endpoint is restricted to localhost or a private network.

Encryption of RPC traffic (for example using TLS) is not part of this ticket but might be considered in the future.

### PR DESCRIPTION
**Pull request by oxarbitrage**
## Motivation

We want to authenticate the RPC method by the zcashd cookie method.

Close https://github.com/ZcashFoundation/zebra/issues/8864

## Solution

- Added `enable_cookie_auth` config field to use cookie authentication, enabled by default.
- Generate cookie when the RPC endpoint starts.
- Start the RPC endpoint server with that generated cookie.
- In the middleware, if auth is enabled, check for basic HTTP auth header.
  - Redirect to request if the user provided auth matches the one generated that the server has.
  - Else error 401.

### Tests

- Manual
- All tests should remain working as we disabled auth for all of them.

### Follow-up Work

- Make the python RPC tests work with authproxy again, or/and:
  - Create a rust test to query with auth. 

### PR Author's Checklist

- [x] The PR name will make sense to users.
- [ ] The PR provides a CHANGELOG summary.
- [x] The solution is tested.
- [ ] The documentation is up to date.
- [x] The PR has a priority label.

### PR Reviewer's Checklist

- [ ] The PR Author's checklist is complete.
- [ ] The PR resolves the issue.

### CODE REVIEW

**Comment by oxarbitrage** [comment]
Requests from a remote host that don't have the cookie generated at startup will be rejected. In zcashd, the `zcash-cli` can be used in a remote host that has the cookie and authenticate. However, we are not targeting a specific application, we want to use the authentication method in a generic way from a remote client. One option can be to use curl with the cookie file and compare server side. Other options are welcome, i am still brainstorming it.

**Comment by oxarbitrage** [comment]
We had a chat today about this with @upbqdn and @arya2, i am adding some more research here. 

About the cookie auth method:

> ... Read access to this file controls who can access through RPC ...

https://bitcoin.org/en/release/v0.12.0#rpc-random-cookie-rpc-authentication

That means the cookie method is actually for local access. I think we should focus on that in this PR.

For remote access, we thought in username/password over TLS/SSL as an option. Bitcoin supported this for its RPC endpoint in the past however they don't do it anymore claiming that the RPC access should only be shared with trusted environments.

https://en.bitcoin.it/wiki/Enabling_SSL_on_original_client_daemon#:~:text=SSL%20for%20RPC%20in%20Bitcoin,was%20criticized%20for%20some%20time.

https://bitcoin.stackexchange.com/questions/108293/why-does-bitcoin-cores-rpc-interface-not-use-encryption

It seems that the remote access should be a combination of username and password, with the additional `rpcallowip` config option plus a vpn layer.

**Comment by oxarbitrage** [comment]
Continuing with the cookie auth method, the zcash-cli sends the cookie content as basic HTTP credentials to the server: https://github.com/zcash/zcash/blob/master/src/bitcoin-cli.cpp#L251
https://github.com/zcash/zcash/blob/master/src/bitcoin-cli.cpp#L266

We want to do that but just with `curl`, we need to intercept the `Authorization` in the zebra RPC middleware and compare it with the cookie in the server side.

I got confused thinking the cookie method will work for remote access, my apologies for that.

**Comment by upbqdn** [comment]
> That means the cookie method is actually for local access. I think we should focus on that in this PR.

Do we have any use cases that require authentication for local access, though?

**Comment by oxarbitrage** [comment]
> Do we have any use cases that require authentication for local access, though?

It's a security measure. You can't access the resources if you don't have read access to the cookie, even if you are in the same machine.

File: zebra-rpc/src/config.rs [hunk 1]
```diff
@@ -94,6 +99,9 @@ impl Default for Config {
 
             // Debug options are always off by default.
             debug_force_finished_sync: false,
+
+            //
```
Comment (upbqdn) [inline zebra-rpc/src/config.rs]:
```suggestion
            // Use the default cache dir for the auth cookie.
```

File: zebra-rpc/src/methods.rs [hunk 2]
```diff
@@ -314,6 +314,10 @@ pub trait Rpc {
     /// tags: control
     #[rpc(name = "stop")]
     fn stop(&self) -> Result<String>;
+
+    #[rpc(name = "unauthenticated")]
+    /// A dummy RPC method that just returns a non-authenticated RPC error.
```
Comment (upbqdn) [inline zebra-rpc/src/methods.rs]: ###### Optional

```suggestion
    /// A dummy RPC method that just returns a non-authenticated RPC error.
    #[rpc(name = "unauthenticated")]
```

File: zebra-rpc/src/server/cookie.rs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    io::{Read, Write},
+    path::PathBuf,
+};
+
+/// The user field in the cookie (arbitrary, only for recognizability in debugging/logging purposes)
+pub const COOKIEAUTH_USER: &str = "__cookie__";
+/// Default name for auth cookie file */
+pub const COOKIEAUTH_FILE: &str = ".cookie";
+
+/// Generate a new auth cookie and return the encoded password.
```
Comment (upbqdn) [inline zebra-rpc/src/server/cookie.rs]:
```suggestion
/// Generate a new auth cookie and store it in the given `cookie_dir`.
```

File: zebra-rpc/src/server/cookie.rs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+pub fn generate(cookie_dir: PathBuf) -> Option<()> {
```
Comment (upbqdn) [inline zebra-rpc/src/server/cookie.rs]: ###### Optional

Returning `Result` would be more expressive.

File: zebra-rpc/src/server/cookie.rs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    let mut file = File::open(cookie_dir.join(COOKIEAUTH_FILE)).ok()?;
+    let mut contents = String::new();
+    file.read_to_string(&mut contents).ok()?;
+
+    let parts: Vec<&str> = contents.split(":").collect();
+    Some(parts[1].to_string())
+}
+
+/// Delete the auth cookie.
+pub fn delete() -> Option<()> {
```
Comment (upbqdn) [inline zebra-rpc/src/server/cookie.rs]: ###### Optional

Similarly here.

File: zebra-rpc/src/server/http_request_compatibility.rs [hunk 6]
```diff
@@ -141,4 +149,51 @@ impl FixHttpRequestMiddleware {
...
+            .and_then(|token| URL_SAFE.decode(token).ok())
+            .and_then(|decoded| String::from_utf8(decoded).ok())
+            .and_then(|decoded_str| {
+                decoded_str
+                    .split(':')
+                    .nth(1)
+                    .map(|password| password.to_string())
+            })
+            .map_or(false, |password| {
+                if let Some(cookie_password) = cookie::get(self.0.cookie_dir.clone()) {
```
Comment (upbqdn) [inline zebra-rpc/src/server/http_request_compatibility.rs]: We read the cookie from the disk on each check here and only rely on the OS caching the disk reads. We could keep the cookie in the memory instead. For example, by using `lazy_static!` in `cookie::get`.

File: zebra-rpc/src/server/http_request_compatibility.rs [hunk 7]
```diff
@@ -141,4 +149,51 @@ impl FixHttpRequestMiddleware {
             );
         }
     }
+
+    /// Change the method name in the JSON request.
```
Comment (upbqdn) [inline zebra-rpc/src/server/http_request_compatibility.rs]: ###### Optional

Could we move the contents of this fn to `unauthenticated` and remove the fn since we call it only from there, and it only uses a single-use, hard-coded method name?

File: zebrad/tests/common/checkpoints.rs [hunk 8]
```diff
@@ -504,7 +505,8 @@ pub fn wait_for_zebra_checkpoints_generation<
 /// Returns an approximate `zebrad` tip height, using JSON-RPC.
 #[tracing::instrument]
 pub async fn zebrad_tip_height(zebra_rpc_address: SocketAddr) -> Result<Height> {
-    let client = RpcRequestClient::new(zebra_rpc_address);
+    let auth_cookie = cookie::get(RpcConfig::default().cookie_dir).expect("cookie should exist");
+    let client = RpcRequestClient::new(zebra_rpc_address, auth_cookie);
```
Comment (upbqdn) [inline zebrad/tests/common/checkpoints.rs]: We write the cookie to the disk, and then read it even in tests that are not related to the authentication. We could generate the cookie for tests without writing it to the disk at all, and instantiate it only through `lazy_static!`.

**Review by upbqdn — DISMISSED** [review]

File: zebra-rpc/src/server/http_request_compatibility.rs [hunk 9]
```diff
@@ -34,13 +37,18 @@ use jsonrpc_http_server::{
...
+#[derive(Clone, Debug)]
+pub struct FixHttpRequestMiddleware(pub crate::config::Config);
 
 impl RequestMiddleware for FixHttpRequestMiddleware {
     fn on_request(&self, mut request: Request<Body>) -> RequestMiddlewareAction {
         tracing::trace!(?request, "original HTTP request");
 
+        // Check if the request is authenticated
+        if !self.check_credentials(request.headers_mut()) {
+            request = Self::unauthenticated(request);
```
Comment (arya2) [inline zebra-rpc/src/server/http_request_compatibility.rs]: This should work for responding here instead of changing the method name (though it doesn't return the request id or jsonrpc version, it could but it would have to parse the request body first, which didn't seem worthwhile):

```suggestion
            let error = jsonrpc_core::Error {
                code: jsonrpc_core::ErrorCode::ServerError(401),
                message: "unauthenticated method".to_string(),
                data: None,
            };

            return jsonrpc_http_server::Response {
                code: jsonrpc_http_server::hyper::StatusCode::from_u16(401)
                    .expect("hard-coded status code should be valid"),
                content_type: header::HeaderValue::from_static("application/json; charset=utf-8"),
                content: serde_json::to_string(&jsonrpc_core::Response::from(error, None))
                    .expect("hard-coded result should serialize"),
            }
            .into();
```

File: zebra-rpc/src/config.rs (hunk 1 above)
Comment (oxarbitrage) [reply zebra-rpc/src/config.rs]: https://github.com/ZcashFoundation/zebra/pull/8900/commits/77082947877c025b51de5d61a99e63769eeb9215

File: zebra-rpc/src/server/cookie.rs (hunk 3 above)
Comment (oxarbitrage) [reply zebra-rpc/src/server/cookie.rs]: Updated.

File: zebra-rpc/src/server/cookie.rs (hunk 4 above)
Comment (oxarbitrage) [reply zebra-rpc/src/server/cookie.rs]: https://github.com/ZcashFoundation/zebra/pull/8900/commits/d48d9d4cdfbbd7edac5cd45819f4c22fc1463b70

File: zebra-rpc/src/server/cookie.rs (hunk 5 above)
Comment (oxarbitrage) [reply zebra-rpc/src/server/cookie.rs]: https://github.com/ZcashFoundation/zebra/pull/8900/commits/d48d9d4cdfbbd7edac5cd45819f4c22fc1463b70

File: zebra-rpc/src/methods.rs (hunk 2 above)
Comment (oxarbitrage) [reply zebra-rpc/src/methods.rs]: Do not apply anymore thanks to https://github.com/ZcashFoundation/zebra/pull/8900/commits/963c1d26348c70fdec92955fd1ebb4077c208355

File: zebra-rpc/src/server/http_request_compatibility.rs (hunk 7 above)
Comment (oxarbitrage) [reply zebra-rpc/src/server/http_request_compatibility.rs]: `unauthenticated` is gone in https://github.com/ZcashFoundation/zebra/pull/8900/commits/963c1d26348c70fdec92955fd1ebb4077c208355

File: zebra-rpc/src/server/http_request_compatibility.rs (hunk 9 above)
Comment (oxarbitrage) [reply zebra-rpc/src/server/http_request_compatibility.rs]: Thanks, that is really good! Added in https://github.com/ZcashFoundation/zebra/pull/8900/commits/963c1d26348c70fdec92955fd1ebb4077c208355

File: zebra-rpc/src/server/http_request_compatibility.rs (hunk 6 above)
Comment (oxarbitrage) [reply zebra-rpc/src/server/http_request_compatibility.rs]: Thanks. I tried that approach to address the read/write issue, but I realized we also want authentication to be enabled by default, with the option to disable it. https://github.com/ZcashFoundation/zebra/issues/8864.

To address this, I added a configuration option to toggle authentication and disabled auth for all tests. 

I manually tested the system, but we should add specific tests to verify the cookie-based authentication method.

File: zebrad/tests/common/checkpoints.rs (hunk 8 above)
Comment (oxarbitrage) [reply zebrad/tests/common/checkpoints.rs]: https://github.com/ZcashFoundation/zebra/pull/8900#discussion_r1798481844

**Comment by mpguerra** [comment]
Can this be merged once approved?

File: zebra-rpc/src/server/cookie.rs [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    let mut contents = String::new();
+    file.read_to_string(&mut contents)?;
+
+    let parts: Vec<&str> = contents.split(":").collect();
+    Ok(parts[1].to_string())
+}
+
+/// Delete the auth cookie.
+pub fn delete() -> Result<()> {
+    remove_file(COOKIEAUTH_FILE)?;
```
Comment (upbqdn) [inline zebra-rpc/src/server/cookie.rs]: I don't think the removal works because we're using only the file name without the path to it.

File: zebra-rpc/src/server/http_request_compatibility.rs [hunk 11]
```diff
@@ -141,4 +161,30 @@ impl FixHttpRequestMiddleware {
...
+            .and_then(|token| URL_SAFE.decode(token).ok())
+            .and_then(|decoded| String::from_utf8(decoded).ok())
+            .and_then(|decoded_str| {
+                decoded_str
+                    .split(':')
+                    .nth(1)
+                    .map(|password| password.to_string())
+            })
+            .map_or(false, |password| {
+                if let Ok(cookie_password) = cookie::get(self.0.cookie_dir.clone()) {
```
Comment (upbqdn) [inline zebra-rpc/src/server/http_request_compatibility.rs]: We're still reading the cookie from the disk on each RPC request.

**Comment by upbqdn** [comment]
> Can this be merged once approved?

Nope.

**Comment by upbqdn** [comment]
This PR is missing a priority label.

File: zebra-rpc/src/server/cookie.rs (hunk 10 above)
Comment (oxarbitrage) [reply zebra-rpc/src/server/cookie.rs]: Fixed in https://github.com/ZcashFoundation/zebra/pull/8900/commits/43131b8a5645f76c0d3b453a0099d9a7061b5f5a

File: zebra-rpc/src/server/http_request_compatibility.rs (hunk 11 above)
Comment (oxarbitrage) [reply zebra-rpc/src/server/http_request_compatibility.rs]: https://github.com/ZcashFoundation/zebra/pull/8900/commits/43131b8a5645f76c0d3b453a0099d9a7061b5f5a

**Review by upbqdn — APPROVED** [review]

**Comment by upbqdn** [comment]
Should we also update user docs and describe where users can find the cookie and how to use it?
